app/src/main/kotlin/compiler/intermediate_form/FunctionDependenciesAnalyzer.kt

```diff
@@ -16,7 +14,95 @@ object FunctionDependenciesAnalyzer {
         return TODO()
     }
 
-    fun createCallGraph(ast: Program): ReferenceMap<Function, ReferenceSet<Function>> {
-        return TODO()
+    fun createCallGraph(ast: Program, nameResolution: ReferenceMap<Any, NamedNode>): ReferenceMap<Function, ReferenceSet<Function>> {
+
+        val functionCalls: ReferenceHashMap<Function, ReferenceSet<Function>> = ReferenceHashMap()
+
+        fun getCalledFunctions(node: Any?): ReferenceSet<Function> {
+
+            fun getCalledOfStatementBlock(statementBlock: StatementBlock?): ReferenceSet<Function> =
+                    if (statementBlock === null) referenceSetOf() else combineReferenceSets(statementBlock.map { getCalledFunctions(it) })
+
+            return when (node) {
+
+                is Program.Global.FunctionDefinition -> {
+                    functionCalls[node.function] = getCalledOfStatementBlock(node.function.body)
+                    return referenceSetOf()
+                }
+
+                // Expressions
+
+                is Expression.FunctionCall -> combineReferenceSets(
+                            referenceSetOf(nameResolution[node] as Function),
+                            combineReferenceSets(node.arguments.map { getCalledFunctions(it.value) }),
+                    )
+
+                is Expression.UnaryOperation -> getCalledFunctions(node.operand)
+
+                is Expression.BinaryOperation -> combineReferenceSets(
+                        getCalledFunctions(node.leftOperand),
+                        getCalledFunctions(node.rightOperand),
+                )
+
+                is Expression.Conditional -> combineReferenceSets(
+                        getCalledFunctions(node.condition),
+                        getCalledFunctions(node.resultWhenTrue),
+                        getCalledFunctions(node.resultWhenFalse),
+                )
+
+                // Statements
+
+                is Statement.Evaluation -> getCalledFunctions(node.expression)
+
+                is Statement.VariableDefinition -> getCalledFunctions(node.variable.value)
+
+                is Statement.FunctionDefinition -> {
+                    functionCalls[node.function] = getCalledOfStatementBlock(node.function.body)
+                    return combineReferenceSets(node.function.parameters.map { it.defaultValue }.map { getCalledFunctions(it) })
```

Comment:
Maybe `node.function.parameters.map { getCalledFunctions(it.defaultValue) }` would be simpler?